llvm/lib/CodeGen/MachineTraceMetrics.cpp

```diff
@@ -761,6 +762,64 @@ static void updatePhysDepsDownwards(const MachineInstr *UseMI,
   }
 }
 
+/// Estimates the number of cycles elapsed between DefMI and UseMI if they're
+/// non-null and in the same BasicBlock. Returns std::nullopt when UseMI is in a
+/// different MBB than DefMI, or when it is a dangling MI.
+static std::optional<unsigned>
+estimateDefUseCycles(const TargetSchedModel &Sched, const MachineInstr *DefMI,
+                     const MachineInstr *UseMI) {
+  if (!DefMI || !UseMI || DefMI == UseMI)
+    return 0;
+  const MachineBasicBlock *ParentBB = DefMI->getParent();
+  if (ParentBB != UseMI->getParent())
+    return std::nullopt;
+
+  const auto DefIt =
+      llvm::find_if(ParentBB->instrs(),
+                    [DefMI](const MachineInstr &MI) { return DefMI == &MI; });
+  const auto UseIt =
+      llvm::find_if(ParentBB->instrs(),
+                    [UseMI](const MachineInstr &MI) { return UseMI == &MI; });
+  assert(std::distance(DefIt, UseIt) > 0 &&
+         "Def expected to appear before use");
+  unsigned NumMicroOps = 0;
+  for (auto It = DefIt; It != UseIt; ++It) {
+    // In some cases, UseMI is a dangling MI beyond the end of the MBB.
+    if (It.isEnd())
+      return std::nullopt;
```

Comment:
When we return `nullopt`, `computeOperandLatency` uses 0 cycles as its estimate. Wouldn't it be better to use the DefaultDefLatency?

Reply:
If the UseMI is a dangling MI beyond the end of the MBB, isn't it as good as no use, and hence an estimate of 0 cycles?

Reply:
```
bb1:
  DefMI = ...
  br bb2
bb2:
  UseMI = ...
```

In this example, UseMI is beyond bb1, but there I'm not sure we can say this behaves like no use. It is executed 2 instructions after DefMI is created.

Reply:
Wouldn't UseMI's parent be bb2, and not bb1 in this case? We already check that DefMI and UseMI have the same parent MBB.

Reply:
Do you mind sharing a minimal example of what it looks like to be a "dangling use"?

Reply:
Thanks for pressing me on this! I initially thought it was an underlying bug where the UseMI is removed from the MBB, but upon further investigation, this is not the case: UseMI is actually a PHI at the beginning of the MBB in this case.

```
bb:
  UseMI = PHI ...
  ...
  DefMI = ...
  ...
  br %bb
```

Will update the patch shortly.